### PR TITLE
[FLINK-9126] New CassandraPojoInputFormat to output data as a custom annotated Cassandra Pojo

### PR DESCRIPTION
**Pull request by bmeriaux**
## What is the purpose of the change

Committing the new CassandraPojoInputFormat class. This works similarly to the CassandraInputClass, but allows the data that is read from Cassandra to be output as a custom POJO that the user has created whichhas been annotated using Datastax API.
## Brief change log

- Initial commit of the CassandraPojoInputFormat class and validation test.
## Verifying this change

- CassandraPojoInputFormat can be validated with the testCassandraBatchPojoFormat test in the CassandraConnectorITCaseTest.java file.
## Does this pull request potentially affect one of the following parts:

- Dependencies (does it add or upgrade a dependency): no
- The public API, i.e., is any changed class annotated with @Public(Evolving): no
- The serializers: no
- The runtime per-record code paths (performance sensitive): don't know
- Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: no
- The S3 file system connector: no

## Documentation

- Does this pull request introduce a new feature? yes
- If yes, how is the feature documented? JavaDocs

### CODE REVIEW

**Comment by bmeriaux** [comment]
continuation of #5876

File: flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraInputFormatBase.java [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+
+import com.datastax.driver.core.Cluster;
+import com.datastax.driver.core.Session;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+
+/**
+ * InputFormat to read data from Apache Cassandra and generate a custom Cassandra annotated object.
```
Comment (azagrebin) [inline flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraInputFormatBase.java]: Comment:
Base class for {@link RichInputFormat} to read data from Apache Cassandra.

File: flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraPojoInputFormat.java [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+ *
+ * @param <OUT> type of inputClass
+ */
+public class CassandraPojoInputFormat<OUT> extends CassandraInputFormatBase<OUT> {
+
+	private static final long serialVersionUID = 1992091320180905115L;
+	private final MapperOptions mapperOptions;
+
+	private transient Result<OUT> resultSet;
+	private Class<OUT> inputClass;
```
Comment (azagrebin) [inline flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraPojoInputFormat.java]: can be `final`

File: flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraPojoInputFormat.java [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+	public CassandraPojoInputFormat(String query, ClusterBuilder builder, Class<OUT> inputClass) {
+		this(query, builder, inputClass, null);
+	}
+
+	public CassandraPojoInputFormat(String query, ClusterBuilder builder, Class<OUT> inputClass, MapperOptions mapperOptions) {
+		super(query, builder);
+		this.mapperOptions = mapperOptions;
+		Preconditions.checkArgument(inputClass != null, "InputClass cannot be null");
```
Comment (azagrebin) [inline flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraPojoInputFormat.java]: `checkNotNull`

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/CustomCassandraAnnotatedPojo.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+ * limitations under the License.
+ */
+
+package org.apache.flink.batch.connectors.cassandra;
+
+import com.datastax.driver.mapping.annotations.Column;
+import com.datastax.driver.mapping.annotations.Table;
+
+/**
+ * Example of Cassandra Annotated POJO class for use with CassandraInputFormatter.
```
Comment (azagrebin) [inline flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/CustomCassandraAnnotatedPojo.java]: In doc comment:
`CassandraInputFormatter` -> `{@link CassandraPojoInputFormat}`

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java [hunk 5]
```diff
@@ -470,6 +472,34 @@ public void testCassandraTableSink() throws Exception {
...
+
+		List<CustomCassandraAnnotatedPojo> result = new ArrayList<>();
+
+		while (!source.reachedEnd()) {
+			CustomCassandraAnnotatedPojo temp = source.nextRecord(new CustomCassandraAnnotatedPojo());
+			result.add(temp);
+		}
+
+		source.close();
+		Assert.assertEquals(20, result.size());
```
Comment (azagrebin) [inline flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java]: Can we also convert `collection` of tuples into list of `CustomCassandraAnnotatedPojo` and check the contents as well, additionally to size?

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/example/BatchPojoExample.java [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+import org.apache.flink.streaming.connectors.cassandra.ClusterBuilder;
+
+import com.datastax.driver.core.Cluster;
+import com.datastax.driver.core.ConsistencyLevel;
+import com.datastax.driver.mapping.Mapper;
+
+import java.util.ArrayList;
+
+/**
+ * This is an example showing the to use the {@link CassandraPojoInputFormat}-/CassandraOutputFormats in the Batch API.
```
Comment (azagrebin) [inline flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/example/BatchPojoExample.java]: does `-/CassandraOutputFormat` mean something?

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/example/BatchPojoExample.java [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+ *
+ * <p>The example assumes that a table exists in a local cassandra database, according to the following queries:
+ * CREATE KEYSPACE IF NOT EXISTS test WITH replication = {'class': 'SimpleStrategy', 'replication_factor': ‘1’};
```
Comment (azagrebin) [inline flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/example/BatchPojoExample.java]: I think there is still some mismatch between queries, doc comments, tuples in `dataSet ` and fields of `CustomCassandraAnnotatedPojo`. `testCassandraBatchPojoFormat` looks better from this side.

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/example/BatchPojoExample.java [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+
+		DataSet<Tuple2<Integer, String>> dataSet = env.fromCollection(collection);
+
+		dataSet.output(new CassandraTupleOutputFormat<Tuple2<Integer, String>>(INSERT_QUERY, new ClusterBuilder() {
+			@Override
+			protected Cluster buildCluster(Cluster.Builder builder) {
+				return builder.addContactPoints("127.0.0.1").build();
+			}
+		}));
+
```
Comment (azagrebin) [inline flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/example/BatchPojoExample.java]: I think `env.execute("Write");` should be here, not at the end.

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/example/BatchPojoExample.java [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+		env.setParallelism(1);
+
+		ArrayList<Tuple2<Integer, String>> collection = new ArrayList<>(20);
+		for (int i = 0; i < 20; i++) {
+			collection.add(new Tuple2<>(i, "string " + i));
+		}
+
+		DataSet<Tuple2<Integer, String>> dataSet = env.fromCollection(collection);
+
+		dataSet.output(new CassandraTupleOutputFormat<Tuple2<Integer, String>>(INSERT_QUERY, new ClusterBuilder() {
```
Comment (azagrebin) [inline flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/example/BatchPojoExample.java]: we can remove `Tuple2<Integer, String>` and use empty `<>` (two places)

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/example/BatchPojoExample.java [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+			protected Cluster buildCluster(Cluster.Builder builder) {
+				return builder.addContactPoints("127.0.0.1").build();
+			}
+		}));
+
+		/*
+		 *	This is for the purpose of showing an example of creating a DataSet using CassandraPojoInputFormat.
+		 */
+		DataSet<CustomCassandraAnnotatedPojo> inputDS = env
+			.createInput(new CassandraPojoInputFormat<>(SELECT_QUERY, new ClusterBuilder() {
```
Comment (azagrebin) [inline flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/example/BatchPojoExample.java]: `ClusterBuilder` should also have `serialVersionUID`. It also duplicated in 2 places. We can create a variable to define it once.

File: flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraInputFormatBase.java [hunk 11]
```diff
@@ -0,0 +1,104 @@
...
+
+	protected final String query;
+	protected final ClusterBuilder builder;
+
+	protected transient Cluster cluster;
+	protected transient Session session;
+
+	public CassandraInputFormatBase(String query, ClusterBuilder builder){
+		Preconditions.checkArgument(!Strings.isNullOrEmpty(query), "Query cannot be null or empty");
+		Preconditions.checkArgument(builder != null, "Builder cannot be null");
```
[the remark this answers is not in the document]
Comment (bmeriaux) [reply flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraInputFormatBase.java]: it will throw a NullPointerException instead of IllegallArgumentException, like others classes of the connector, i think we should keep checkArgument

File: flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraPojoInputFormat.java (hunk 3 above)
Comment (bmeriaux) [reply flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraPojoInputFormat.java]: like previous one, i think we should stick on checkArgument to keep a IllegalArgumentException

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java (hunk 5 above)
Comment (bmeriaux) [reply flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java]: To save easily a list of CustomCassandraAnnotatedPojo, i used the CassandraPojoSink

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/example/BatchPojoExample.java (hunk 8 above)
Comment (azagrebin) [reply flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/example/BatchPojoExample.java]: I think it should be first creating output then executing:
```
dataSet.output(new CassandraTupleOutputFormat<>(INSERT_QUERY, clusterBuilder));
env.execute("Write");
```

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java (hunk 5 above)
Comment (azagrebin) [reply flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java]: maybe example would be also more consistent if we use `CassandraPojoSink` there as well to populate the table, instead of raw tuples.

File: flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraInputFormatBase.java [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+/**
+ * Base class for {@link RichInputFormat} read data from Apache Cassandra and generate a custom Cassandra annotated object.
+ * @param <OUT> type of inputClass
+ */
+abstract class CassandraInputFormatBase<OUT> extends RichInputFormat<OUT, InputSplit> implements NonParallelInput {
+	private static final Logger LOG = LoggerFactory.getLogger(CassandraInputFormatBase.class);
+	private static final long serialVersionUID = -1519372881115104601L;
+
+	final String query;
+	final ClusterBuilder builder;
```
Comment (azagrebin) [inline flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraInputFormatBase.java]: can be private

File: flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraInputFormatBase.java [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+
+	transient Cluster cluster;
+	transient Session session;
+
+	public CassandraInputFormatBase(String query, ClusterBuilder builder){
```
Comment (azagrebin) [inline flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraInputFormatBase.java]: can be also package-private

File: flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraInputFormat.java [hunk 14]
```diff
@@ -44,90 +31,38 @@
...
-	public BaseStatistics getStatistics(BaseStatistics cachedStatistics) throws IOException {
-		return cachedStatistics;
+		super(query, builder);
 	}
 
 	/**
 	 * Opens a Session and executes the query.
 	 *
-	 * @param ignored
+	 * @param ignored because parameter is not parallelizable.
```
Comment (azagrebin) [inline flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraInputFormat.java]: the exception can also go from the doc string

File: flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraInputFormatBase.java (hunk 11 above)
[the remark this answers is not in the document]
Comment (azagrebin) [reply flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraInputFormatBase.java]: Well, I think we consistently use `checkNotNull` mostly in Flink, also java util packages throw `NullPointerException` for unsupported `null` args.

**Comment by bmeriaux** [comment]
I think it is ready now :), should I rebase on master ?

**Comment by azagrebin** [comment]
Great, looks good! I think committer will do it but rebasing on master is always a good thing.

The `CassandraPojoSink` is actually for streaming API. Unless we want to extend `CassandraOutputFormatBase` with `CassandraPojoOutputFormat`, I think it is better to use `CassandraTupleOutputFormat` for batch example in this PR, as it was. 

Sorry for confusion, can we roll it back to `CassandraTupleOutputFormat` for batch example before merge (with `execute("write")` after `output`)? `CassandraOutputFormatBase` can be extended with `CassandraPojoOutputFormat` in other PR if needed.

**Comment by bmeriaux** [comment]
ok, no problem, it will be a little more complex than just to extend `CassandraOutputFormatBase` to do `CassandraPojoOutputFormat`, but i could do it later in another PR

**Comment by bmeriaux** [comment]
ok, thanks, i just repush to change my email commit author field :)

**Comment by bmeriaux** [comment]
I have created an issue for `CassandraPojoOutputFormat` and have the code for it, it will wait for the merge of this one :)

File: flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraInputFormatBase.java [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+import com.datastax.driver.core.Session;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Base class for {@link RichInputFormat} read data from Apache Cassandra and generate a custom Cassandra annotated object.
+ * @param <OUT> type of inputClass
+ */
+abstract class CassandraInputFormatBase<OUT> extends RichInputFormat<OUT, InputSplit> implements NonParallelInput {
+	private static final Logger LOG = LoggerFactory.getLogger(CassandraInputFormatBase.class);
```
Comment (zentol) [inline flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraInputFormatBase.java]: Let's make this non-static and initialize the logger with `LoggerFactory.getLogger(getClass())`; then sub-classes don't need to create their own logger, we can have a single source for logging messages, and the class is the actually user-facing one (which is what one would expect).

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/CustomCassandraAnnotatedPojo.java [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+import com.datastax.driver.mapping.annotations.Column;
+import com.datastax.driver.mapping.annotations.Table;
+
+/**
+ * Example of Cassandra Annotated POJO class for use with {@link CassandraPojoInputFormat}.
+ */
+@Table(name = "batches", keyspace = "flink")
+public class CustomCassandraAnnotatedPojo {
+
+
```
Comment (zentol) [inline flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/CustomCassandraAnnotatedPojo.java]: double empty-lines should be prohibited by checkstyle, don't know why this isn't failing :/

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/example/BatchPojoExample.java [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+
+		dataSet.output(new CassandraTupleOutputFormat<>(INSERT_QUERY, clusterBuilder));
+
+		env.execute("Write");
+
+		/*
+		 *	This is for the purpose of showing an example of creating a DataSet using CassandraPojoInputFormat.
+		 */
+		DataSet<CustomCassandraAnnotatedPojo> inputDS = env
+			.createInput(new CassandraPojoInputFormat<>(SELECT_QUERY, clusterBuilder, CustomCassandraAnnotatedPojo.class, () -> new Mapper.Option[]{Mapper.Option.consistencyLevel(ConsistencyLevel.ANY)}));
```
Comment (zentol) [inline flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/batch/connectors/cassandra/example/BatchPojoExample.java]: place arguments on separate lines for readability

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java [hunk 18]
```diff
@@ -470,6 +477,41 @@ public void testCassandraTableSink() throws Exception {
 		Assert.assertTrue("The input data was not completely written to Cassandra", input.isEmpty());
 	}
 
+	@Test
+	public void testCassandraBatchPojoFormat() throws Exception {
+
+		session.execute(CREATE_TABLE_QUERY.replace(TABLE_NAME_VARIABLE, "batches"));
```
Comment (zentol) [inline flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java]: This shouldn't be necessary; the test infrastructure already creates a table.

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java [hunk 19]
```diff
@@ -470,6 +477,41 @@ public void testCassandraTableSink() throws Exception {
...
+
+		CassandraPojoSink<CustomCassandraAnnotatedPojo> sink = new CassandraPojoSink<>(CustomCassandraAnnotatedPojo.class, builder);
+		sink.open(new Configuration());
+
+		List<CustomCassandraAnnotatedPojo> customCassandraAnnotatedPojos = IntStream.range(0, 20)
+			.mapToObj(x -> new CustomCassandraAnnotatedPojo(UUID.randomUUID().toString(), x, 0))
+			.collect(Collectors.toList());
+
+		customCassandraAnnotatedPojos.forEach(sink::send);
+		sink.close();
```
Comment (zentol) [inline flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java]: this should be done in a `finally` block

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java [hunk 20]
```diff
@@ -470,6 +477,41 @@ public void testCassandraTableSink() throws Exception {
...
+		source.configure(new Configuration());
+		source.open(null);
+
+		List<CustomCassandraAnnotatedPojo> result = new ArrayList<>();
+		while (!source.reachedEnd()) {
+			CustomCassandraAnnotatedPojo temp = source.nextRecord(null);
+			result.add(temp);
+		}
+
+		source.close();
```
Comment (zentol) [inline flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java]: this should be done in a `finally` block

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java [hunk 21]
```diff
@@ -470,6 +477,41 @@ public void testCassandraTableSink() throws Exception {
 		Assert.assertTrue("The input data was not completely written to Cassandra", input.isEmpty());
 	}
 
+	@Test
+	public void testCassandraBatchPojoFormat() throws Exception {
+
+		session.execute(CREATE_TABLE_QUERY.replace(TABLE_NAME_VARIABLE, "batches"));
+
+		CassandraPojoSink<CustomCassandraAnnotatedPojo> sink = new CassandraPojoSink<>(CustomCassandraAnnotatedPojo.class, builder);
```
Comment (zentol) [inline flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java]: I'm a bit torn on this one. I kinda dislike that we're using other sinks for the verification; this test could now fail if the `CassandraPojoSink` is modified in some incompatible way, which simply shouldn't be possible.

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java [hunk 22]
```diff
@@ -470,6 +477,41 @@ public void testCassandraTableSink() throws Exception {
...
+		sink.open(new Configuration());
+
+		List<CustomCassandraAnnotatedPojo> customCassandraAnnotatedPojos = IntStream.range(0, 20)
+			.mapToObj(x -> new CustomCassandraAnnotatedPojo(UUID.randomUUID().toString(), x, 0))
+			.collect(Collectors.toList());
+
+		customCassandraAnnotatedPojos.forEach(sink::send);
+		sink.close();
+
+		InputFormat<CustomCassandraAnnotatedPojo, InputSplit> source = new CassandraPojoInputFormat<>(SELECT_DATA_QUERY.replace(TABLE_NAME_VARIABLE, "batches"), builder, CustomCassandraAnnotatedPojo.class);
```
Comment (zentol) [inline flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java]: let's add a sanity check that the table actually contains 20 records to guard against errors during the writing.

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java [hunk 23]
```diff
@@ -470,6 +477,41 @@ public void testCassandraTableSink() throws Exception {
...
+			result.add(temp);
+		}
+
+		source.close();
+		Assert.assertEquals(20, result.size());
+		result.sort(Comparator.comparingInt(CustomCassandraAnnotatedPojo::getCounter));
+		customCassandraAnnotatedPojos.sort(Comparator.comparingInt(CustomCassandraAnnotatedPojo::getCounter));
+
+		for (int i = 0; i < result.size(); i++) {
+			assertThat(result.get(i), samePropertyValuesAs(customCassandraAnnotatedPojos.get(i)));
```
Comment (zentol) [inline flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java]: since they are sorted already I believe we could just compare the lists directly.

**Comment by azagrebin** [comment]
Hi @bmeriaux, we can merge it after @zentol's comments are resolved. We usually merge new features or functionality into the new releases unless there is good reason to port to bug fix releases. I suggest to update to 1.7 in future to use new Cassandra connectors. There should be good compatibility between 1.6 and 1.7. We can discuss it if there is any problem with not having Cassandra updates in 1.6.

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java (hunk 18 above)
Comment (bmeriaux) [reply flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java]: this is necessary because i used the sink with a pojo, and the table used is the one defined by annotation

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java (hunk 22 above)
Comment (bmeriaux) [reply flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java]: ok, it may be a little paranoid nop ? :)

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java (hunk 23 above)
Comment (bmeriaux) [reply flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java]: it's a reflex for me to explicitly sort collection before assertions, i think it is more explicit/easier to understand, but i can remove it if it is not in standard :)

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java (hunk 18 above)
Comment (zentol) [reply flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java]: 👍

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java (hunk 23 above)
Comment (zentol) [reply flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java]: its fine to sort them, but you don't need to iterate over the lists.

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java [hunk 24]
```diff
@@ -382,31 +383,31 @@ public void testCassandraCommitter() throws Exception {
 	@Test
 	public void testCassandraTupleAtLeastOnceSink() throws Exception {
 		CassandraTupleSink<Tuple3<String, Integer, Integer>> sink = new CassandraTupleSink<>(injectTableName(INSERT_DATA_QUERY), builder);
-
-		sink.open(new Configuration());
-
-		for (Tuple3<String, Integer, Integer> value : collection) {
-			sink.send(value);
+		try {
```
Comment (zentol) [inline flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java]: please limit your changes to your sink, or move these into a separate commit.

File: flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraOutputFormatBase.java [hunk 25]
```diff
@@ -41,7 +41,7 @@
  * @param <OUT> Type of the elements to write.
  */
 public abstract class CassandraOutputFormatBase<OUT> extends RichOutputFormat<OUT> {
-	private static final Logger LOG = LoggerFactory.getLogger(CassandraOutputFormatBase.class);
+	private final Logger logger = LoggerFactory.getLogger(getClass());
```
Comment (zentol) [inline flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraOutputFormatBase.java]: Please revert changes to this class; this PR is for the InputFormat and shouldn't touch other classes unless absolutely necessary.

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java (hunk 21 above)
Comment (zentol) [reply flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java]: Let's keep it for now but update it once the outputformat is in.

File: flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraOutputFormatBase.java (hunk 25 above)
Comment (bmeriaux) [reply flink-connectors/flink-connector-cassandra/src/main/java/org/apache/flink/batch/connectors/cassandra/CassandraOutputFormatBase.java]: ok

File: flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java (hunk 24 above)
Comment (bmeriaux) [reply flink-connectors/flink-connector-cassandra/src/test/java/org/apache/flink/streaming/connectors/cassandra/CassandraConnectorITCase.java]: ok
